From b3ab31d79bc990ea1fff516ea418b223492b95b2 Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Sun, 25 Aug 2024 09:32:01 +0200
Subject: [PATCH 01/10] Rebrand

---
 build.gradle | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/build.gradle b/build.gradle
index aeb01c42..172d3de0 100644
--- a/build.gradle
+++ b/build.gradle
@@ -12,7 +12,7 @@ compileTestJava.options.encoding = "UTF-8"
 
 java {
     toolchain {
-        languageVersion = JavaLanguageVersion.of(17)
+        languageVersion = JavaLanguageVersion.of(21)
     }
 }
 
@@ -35,11 +35,12 @@ repositories {
 
         // As of Gradle 5.1, you can limit this to only those
         // dependencies you expect from it
-        content {
+        /*content {
             includeGroup 'org.bukkit'
             includeGroup 'org.spigotmc'
-        }
+        }*/
     }
+    maven { url = "https://repo.papermc.io/repository/maven-public/" }
 
     mavenCentral()
     jcenter()
@@ -81,7 +82,7 @@ repositories {
 
 dependencies {
     //Spigot API
-    compileOnly "org.spigotmc:spigot-api:1.21-R0.1-SNAPSHOT"
+    compileOnly "io.papermc.paper:paper-api:1.21-R0.1-SNAPSHOT"
     //WorldGuard API
     compileOnly group: 'com.sk89q.worldguard', name: 'worldguard-bukkit', version: '7.0.7'
     //Vault
-- 
2.45.1.windows.1

