From 7923f07eb9bca42b260a7816739d3d7626c3158e Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Sat, 24 Aug 2024 18:28:31 +0200
Subject: [PATCH 07/11] Add EliteEntity EntityScheduler

---
 .../com/magmaguy/elitemobs/mobconstructor/EliteEntity.java | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/main/java/com/magmaguy/elitemobs/mobconstructor/EliteEntity.java b/src/main/java/com/magmaguy/elitemobs/mobconstructor/EliteEntity.java
index 12e16270..a8e97743 100644
--- a/src/main/java/com/magmaguy/elitemobs/mobconstructor/EliteEntity.java
+++ b/src/main/java/com/magmaguy/elitemobs/mobconstructor/EliteEntity.java
@@ -24,6 +24,8 @@ import com.magmaguy.elitemobs.utils.EventCaller;
 import com.magmaguy.elitemobs.utils.FoliaRunnable;
 import com.magmaguy.magmacore.util.ChatColorConverter;
 import com.magmaguy.magmacore.util.Logger;
+import io.papermc.paper.threadedregions.scheduler.EntityScheduler;
+import io.papermc.paper.threadedregions.scheduler.ScheduledTask;
 import lombok.Getter;
 import lombok.Setter;
 import org.bukkit.Location;
@@ -697,4 +699,9 @@ public class EliteEntity {
         customMetadata.removeAll(tags);
     }
 
+
+    public EntityScheduler getEntityScheduler() {
+        if (livingEntity == null) return null;
+        return livingEntity.getScheduler();
+    }
 }
-- 
2.45.1.windows.1

