From d62d860c682f7e09c03d1d679c977bbe9a83e3e1 Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Sun, 25 Aug 2024 09:32:04 +0200
Subject: [PATCH 08/10] Fix spawn mobs

---
 .../mobconstructor/custombosses/CustomBossEntity.java         | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/main/java/com/magmaguy/elitemobs/mobconstructor/custombosses/CustomBossEntity.java b/src/main/java/com/magmaguy/elitemobs/mobconstructor/custombosses/CustomBossEntity.java
index 22c03760..b9c99912 100644
--- a/src/main/java/com/magmaguy/elitemobs/mobconstructor/custombosses/CustomBossEntity.java
+++ b/src/main/java/com/magmaguy/elitemobs/mobconstructor/custombosses/CustomBossEntity.java
@@ -394,12 +394,12 @@ public class CustomBossEntity extends EliteEntity implements Listener, Persisten
             @Override
             public void run() {
                 if (customBossEntity.isInCombat()) return;
-                if (!customBossEntity.isValid() && !customBossEntity.isPersistent) {
+                if (/*!customBossEntity.isValid() &&*/ !customBossEntity.isPersistent) {
                     cancel();
                     return;
                 }
                 //for unloaded but persistent bosses
-                if (customBossEntity.getLivingEntity() == null || !customBossEntity.isValid()) {
+                if (customBossEntity.getLivingEntity() == null /*|| !customBossEntity.isValid()*/) {
                     return;
                 }
                 int currentLevel = customBossEntity.getLevel();
-- 
2.45.1.windows.1

